File: temperament4fft.hs
import Data.Ratio
import Data.List (intercalate)
import System.Environment (getArgs)

-- given e.g. [e2,e3,e5] this function returns 3^^e3 * 5^^e5 / 2^^e2
p2r :: [Integer] -> Rational
p2r xs = p2r' 1 primes xs
    where primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
          p2r' v [] _ = error "Too long list" -- if prime list became empty, xs is too long
          p2r' v _ [] = v
          p2r' v (p:ps) (x:xs) = p2r' (v * (exp p x)) ps xs
              where exp 2 x = (1 % 2) ^^ x
                    exp p x = (p % 1) ^^ x

-- given n (semitone) and list [e2, e3, e5] it compares n and 3^^e3 * 5^^e5 / 2^^e2
centsdiff :: Double -> [Integer] -> Double
centsdiff n xs = (*100) . (+(-n)) . (*12) . logBase 2 . fromRational . p2r $ xs

-- exponent range for a given prime base
exponentRange :: Integer -> [Integer]
exponentRange base = [0..(ceiling . (*10) . logBase (fromInteger base) $ (2 :: Double))]

-- list of ratios within threshold cents for a given semitone
lstRatios :: [Integer] -> Double -> Integer -> [(Rational, Integer)]
lstRatios usePrimes threshold n' =
    [(p2r xs, round . centsdiff n $ xs)
    | xs <- candidates,
    (<threshold) . abs . centsdiff n $ xs]
    where n = fromInteger n'
          candidates = sequence (map exponentRange usePrimes)

-- format a single approximation: "≒ 135/256 (-8 cents)"
fmtApprox :: (Rational, Integer) -> String
fmtApprox (r, c) = "≒ " ++ show (numerator r) ++ "/" ++ show (denominator r)
                    ++ " (" ++ showSigned c ++ " cents)"
    where showSigned x | x >= 0    = "+" ++ show x
                       | otherwise = show x

-- format a line: "+3 semitone\t≒ 75/64 (-25 cents)\t≒ 625/512 (+45 cents)"
fmtLine :: [Integer] -> Double -> Integer -> String
fmtLine usePrimes threshold n =
    showSigned n ++ " semitone\t"
    ++ intercalate "\t" (map fmtApprox (lstRatios usePrimes threshold n))
    where showSigned x | x >= 0    = "+" ++ show x
                       | otherwise = show x

-- parse command line arguments
parseArgs :: [String] -> ([Integer], Double)
parseArgs args = (parsePrimes args, parseThreshold args)
    where
        parsePrimes ("--primes":val:_) = map read (splitOn ',' val)
        parsePrimes (_:rest) = parsePrimes rest
        parsePrimes [] = [2, 3, 5]

        parseThreshold ("--threshold":val:_) = read val
        parseThreshold (_:rest) = parseThreshold rest
        parseThreshold [] = 50

        splitOn _ [] = []
        splitOn sep str = let (w, rest) = break (== sep) str
                          in w : case rest of
                                     [] -> []
                                     (_:xs) -> splitOn sep xs

-- main
main :: IO ()
main = do
    args <- getArgs
    let (usePrimes, threshold) = parseArgs args
    mapM_ (putStrLn . fmtLine usePrimes threshold) [-11..11]

File: temperament4fft.py
"""
temperament4fft.py

Purpose:
    Find rational approximations r = 2^(-a) * 3^b * 5^c
    for each semitone interval 2^(n/12) in 12-tone equal temperament.
    Useful for choosing FFT sizes composed only of small prime factors.

Run:
    python temperament4fft.py
"""

import math
from fractions import Fraction
from itertools import product


# Primes used for approximation
PRIMES = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]


def p2r(xs):
    """Given exponent list [e2, e3, e5, ...], return (1/2)^e2 * 3^e3 * 5^e5 * ..."""
    v = Fraction(1)
    for p, x in zip(PRIMES, xs):
        if p == 2:
            v *= Fraction(1, 2) ** x
        else:
            v *= Fraction(p) ** x
    return v


def centsdiff(n, xs):
    """Difference in cents between semitone n and the ratio given by exponent list xs."""
    return (12 * math.log2(float(p2r(xs))) - n) * 100


def lst_ratios(n):
    """Find all rational approximations within 50 cents for semitone n."""
    exponent_range = lambda base: range(0, math.ceil(10 * math.log(2, base)) + 1)
    candidates = product(exponent_range(2), exponent_range(3), exponent_range(5))
    return [
        (p2r(xs), round(centsdiff(n, xs)))
        for xs in candidates
        if abs(centsdiff(n, xs)) < 50
    ]


if __name__ == "__main__":
    for n in range(-11, 12):
        approxs = lst_ratios(n)
        entries = "\t".join(
            f"≒ {r.numerator}/{r.denominator} ({cents:+d} cents)"
            for r, cents in approxs
        )
        print(f"{n:+d} semitone\t{entries}")
